mirrord-layer/src/sockets.rs

```diff
@@ -0,0 +1,222 @@
+use std::{
+    collections::HashMap,
+    io::{Read, Write},
+    net::SocketAddr,
+    os::unix::io::AsRawFd,
+    sync::Mutex,
+};
+
+use multi_map::MultiMap;
+use queues::{IsQueue, Queue};
+use socketpair::{socketpair_stream, SocketpairStream};
+use tracing::{debug, error};
+
+pub struct Socket {
+    pub read_fd: i32,
+    pub read_socket: SocketpairStream,
+    pub write_socket: SocketpairStream,
+}
+
+pub struct ConnectionSocket {
+    pub read_fd: i32,
+    pub read_socket: SocketpairStream,
+    pub write_socket: SocketpairStream,
+    pub address: SocketAddr,
+    pub state: ConnectionState,
+}
+
+#[derive(PartialEq)]
+pub enum ConnectionState {
+    Bound,
+    Listening,
+}
+
+pub struct DataSocket {
+    pub connection_id: u16,
+    #[allow(dead_code)]
+    pub read_socket: SocketpairStream, /* Though this is unread, it's necessary to keep the
+                                        * socket open */
+    pub write_socket: SocketpairStream,
+    pub address: SocketAddr,
+}
+
+pub struct Sockets {
+    new_sockets: Mutex<HashMap<i32, Socket>>,
+    connections: Mutex<MultiMap<i32, u16, ConnectionSocket>>, // Keys are sockfd, port
```

Comment:
You can define types so code will be more readable even without docs - 
`type SockFd = i32` `type Port = u16`